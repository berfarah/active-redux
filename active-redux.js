import _Model from './model';
export { _Model as Model };
import _Registry from './registry';
export { _Registry as Registry };
import * as _Attr from './attributes';
export { _Attr as Attr };

export * from './redux/local';
export * from './redux/remote';
