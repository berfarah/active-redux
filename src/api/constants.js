// API
export const API_CONFIGURE = 'AR_API_CONFIGURE';

// REMOTE
export const REMOTE_CLEAR = 'AR_REMOTE_CLEAR';
export const REMOTE_HYDRATE = 'REMOTE_HYDRATE';

export const REMOTE_WILL_READ = 'AR_REMOTE_WILL_READ';
export const REMOTE_READ_DONE = 'AR_REMOTE_READ_DONE';
export const REMOTE_READ_FAILED = 'AR_REMOTE_READ_FAILED';

export const REMOTE_WILL_CREATE = 'AR_REMOTE_WILL_CREATE';
export const REMOTE_CREATE_DONE = 'AR_REMOTE_CREATE_DONE';
export const REMOTE_CREATE_FAILED = 'AR_REMOTE_CREATE_FAILED';

export const REMOTE_WILL_UPDATE = 'AR_REMOTE_WILL_UPDATE';
export const REMOTE_UPDATE_DONE = 'AR_REMOTE_UPDATE_DONE';
export const REMOTE_UPDATE_FAILED = 'AR_REMOTE_UPDATE_FAILED';

export const REMOTE_WILL_DELETE = 'AR_REMOTE_WILL_DELETE';
export const REMOTE_DELETE_DONE = 'AR_REMOTE_DELETE_DONE';
export const REMOTE_DELETE_FAILED = 'AR_REMOTE_DELETE_FAILED';
