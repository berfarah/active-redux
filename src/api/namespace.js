/**
 * @private
 */
class Namespace {
  constructor(value) { this.namespace = value; }
  get value() { return this.namespace; }
  set change(value) { this.namespace = value; }
}

/**
 * Our default namespace for our reducer is 'api'
 * @example
 * import { combineReducers } from 'redux';
 * import { api as activeReduxApi } from 'active-redux';
 *
 * const reducers = combineReducers({ api: activeReduxApi.reducer });
 */
export default new Namespace('api');
