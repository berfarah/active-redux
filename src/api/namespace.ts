/**
 * @private
 */
class Namespace {
  namespace: string;
  constructor(value) { this.namespace = value; }
  get value() { return this.namespace; }
  set change(value) { this.namespace = value; }
}

/**
 * Our default namespace for our reducer is 'api'
 * @example
 * import { combineReducers } from 'redux';
 * import { api as activeReduxApi } from 'active-redux';
 *
 * const reducers = combineReducers({ api: activeReduxApi.reducer });
 */
export default new Namespace('api');
