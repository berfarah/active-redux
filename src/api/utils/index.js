import imm from 'object-path-immutable';

export apiClient from './api-client';

export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

export const resourcesArray = (data) => (
  Array.isArray(data) ? data : [data]
);

export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    // if we don't do this and the ID is a number, it'll create an array
    newState.set(dataObj.type, state[dataObj.type] || {});
    newState.set(`${dataObj.type}.${dataObj.id}`, dataObj);
  });

  return newState.value();
};

export const clearResources = (state, data) => {
  switch (typeof data) {
    case 'object': {
      if (Array.isArray(data)) return data.reduce(clearResources, state);
      const { type, id } = data;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', data], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

export const createAction = (type) => (payload) => ({ type, payload });

export const createReducer = (mappings, initialState) => (state = initialState, action) => (
  mappings[action.type] ? mappings[action.type](state, action) : state
);

export const hasProperty = (object, ...keys) => (
  keys.reduce((acc, key) => {
    if (acc === undefined) return acc;
    return acc[key];
  }, object)
);

export const markPendingResources = (state, { data }) => (
  resourcesArray(data).reduce((acc, resource) => {
    if (!hasProperty(state, 'resources', resource.type, resource.id)) return acc;
    return acc.set(['resources', resource.type, resource.id, 'isPending'], true);
  }, imm(state)).value()
);
