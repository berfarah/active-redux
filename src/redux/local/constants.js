export const LOCAL_CLEAR = 'AR_LOCAL_CLEAR';
export const LOCAL_HYDRATE = 'AR_LOCAL_HYDRATE';
export const LOCAL_READ = 'AR_LOCAL_READ';
export const LOCAL_CREATE = 'AR_LOCAL_CREATE';
export const LOCAL_UPDATE = 'AR_LOCAL_UPDATE';
export const LOCAL_DELETE = 'AR_LOCAL_DELETE';
