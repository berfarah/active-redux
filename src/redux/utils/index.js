import imm from 'object-path-immutable';

export apiClient from './api-client';

export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

export const resourcesArray = (data) => (
  Array.isArray(data) ? data : [data]
);

export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    // if we don't do this and the ID is a number, it'll create an array
    newState.set(dataObj.type, state[dataObj.type] || {});
    newState.set(`${dataObj.type}.${dataObj.id}`, dataObj);
  });

  return newState.value();
};

export const clearResources = (state, resources) => {
  switch (typeof resources) {
    case 'object': {
      if (Array.isArray(resources)) return resources.reduce(clearResources, state);
      const { type, id } = resources;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', resources], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

export const createAction = (type) => (payload) => ({ type, payload });

export const createReducer = (mappings, initialState) => (state = initialState, action) => (
  mappings[action.type] ? mappings[action.type](state, action) : state
);
