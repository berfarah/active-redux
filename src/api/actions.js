import * as ActionTypes from './constants';
import { createAction, apiClient } from './utils';

const wrappedApiRequest = ({
  resource,
  options,
  endpoint,
  dispatch,
  success,
  failure,
}) => apiClient(endpoint, options).then((json) => {
  dispatch(success(json));
  return Promise.resolve(json);
}).catch((error) => {
  const err = error;
  err.resource = resource;

  dispatch(failure(err));
  return Promise.reject(err);
});

export const apiConfigure = createAction(ActionTypes.API_CONFIGURE);
export const remoteClear = createAction(ActionTypes.REMOTE_CLEAR);
export const remoteHydrate = createAction(ActionTypes.REMOTE_HYDRATE);

const remoteWillCreate = createAction(ActionTypes.REMOTE_WILL_CREATE);
const remoteCreateDone = createAction(ActionTypes.REMOTE_CREATE_DONE);
const remoteCreateFailed = createAction(ActionTypes.REMOTE_CREATE_FAILED);
export const remoteCreate = ({ resource, endpoint = resource.type }) => (dispatch, getState) => {
  dispatch(remoteWillCreate(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().remote.apiConfig,
      method: 'POST',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: remoteCreateDone,
    failure: remoteCreateFailed,
  });
};

const remoteWillRead = createAction(ActionTypes.REMOTE_WILL_READ);
const remoteReadDone = createAction(ActionTypes.REMOTE_READ_DONE);
const remoteReadFailed = createAction(ActionTypes.REMOTE_READ_FAILED);
export const remoteRead = ({ resource, endpoint = resource.type }) => (dispatch, getState) => {
  dispatch(remoteWillRead(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().remote.apiConfig,
      method: 'GET',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: remoteReadDone,
    failure: remoteReadFailed,
  });
};

const remoteWillUpdate = createAction(ActionTypes.REMOTE_WILL_UPDATE);
const remoteUpdateDone = createAction(ActionTypes.REMOTE_UPDATE_DONE);
const remoteUpdateFailed = createAction(ActionTypes.REMOTE_UPDATE_FAILED);
export const remoteUpdate = ({ resource, endpoint = `${resource.type}${resource.id}` }) => (
  (dispatch, getState) => {
    dispatch(remoteWillUpdate(resource));
    return wrappedApiRequest({
      resource,
      options: {
        ...getState().remote.apiConfig,
        method: 'PATCH',
        data: { data: resource }
      },
      endpoint,
      dispatch,
      success: remoteUpdateDone,
      failure: remoteUpdateFailed,
    });
  }
);

const remoteWillDelete = createAction(ActionTypes.REMOTE_WILL_DELETE);
const remoteDeleteDone = createAction(ActionTypes.REMOTE_DELETE_DONE);
const remoteDeleteFailed = createAction(ActionTypes.REMOTE_DELETE_FAILED);
export const remoteDelete = ({ resource, endpoint = `${resource.type}${resource.id}` }) => (
  (dispatch, getState) => {
    const options = {
      ...getState().remote.apiConfig,
      method: 'DELETE',
    };

    dispatch(remoteWillDelete(resource));
    return apiClient(endpoint, options).then((json) => {
      dispatch(remoteDeleteDone(resource));
      return Promise.resolve(json);
    }).catch((error) => {
      const err = error;
      err.resource = resource;

      dispatch(remoteDeleteFailed(err));
      return Promise.reject(err);
    });
  }
);
