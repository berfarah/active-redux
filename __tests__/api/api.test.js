import * as subject from 'active-redux/api';
import * as ActionTypes from 'active-redux/api/constants';
import { createAction } from 'active-redux/api/utils';

describe('reducer', () => {
  const person = {
    type: 'people',
    id: 5,
    attributes: {
      hair: 'black',
      age: 25,
    }
  };

  const fullStore = {
    apiConfig: {},
    isCreating: 0,
    isDeleting: 0,
    isReading: 0,
    isUpdating: 0,
    resources: {
      [person.type]: {
        [person.id]: { attributes: { name: 'Johnny' } },
        2: { attributes: { name: 'Yangster' } }
      }
    }
  };

  describe('initialState', () => {
    it('has an empty resources object', () => {
      expect(subject.reducer(undefined, {})).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    const reduce = subject.reducer;

    describe('apiConfigure()', () => {
      it('updates the apiConfig key', () => {
        const state = reduce(undefined, subject.apiConfigure({ baseUrl: 'http://coverhound.com/api' }));
        expect(state).toMatchSnapshot();
      });
    });


    describe('apiHydrate()', () => {
      it('populates the state with resources', () => {
        const state = reduce(undefined, subject.apiHydrate({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiCreate()', () => {
      let state = fullStore;
      const apiWillCreate = createAction(ActionTypes.API_WILL_CREATE);
      const apiCreateDone = createAction(ActionTypes.API_CREATE_DONE);

      it('increments isCreating', () => {
        state = reduce(undefined, apiWillCreate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, apiCreateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiRead()', () => {
      let state = fullStore;
      const apiWillRead = createAction(ActionTypes.API_WILL_READ);
      const apiReadDone = createAction(ActionTypes.API_READ_DONE);

      it('increments isReading', () => {
        state = reduce(undefined, apiWillRead({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, apiReadDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiClear()', () => {
      it('clears the resource', () => {
        const state = reduce(fullStore, subject.apiClear(person.type));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiUpdate()', () => {
      let state = fullStore;
      const apiWillUpdate = createAction(ActionTypes.API_WILL_UPDATE);
      const apiUpdateDone = createAction(ActionTypes.API_UPDATE_DONE);

      it('increments isUpdating and marks the resource as pending', () => {
        state = reduce(state, apiWillUpdate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('updates the state with resources', () => {
        state = reduce(state, apiUpdateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiDelete()', () => {
      let state = fullStore;
      const apiWillDelete = createAction(ActionTypes.API_WILL_DELETE);
      const apiDeleteDone = createAction(ActionTypes.API_DELETE_DONE);

      it('increments isDeleting and marks the resource as pending', () => {
        state = reduce(state, apiWillDelete({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('removes the resources from the store', () => {
        state = reduce(state, apiDeleteDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });
  });
});
