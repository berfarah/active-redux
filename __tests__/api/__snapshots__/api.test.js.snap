// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer actions apiConfigure() updates the apiConfig key 1`] = `
Object {
  "apiConfig": Object {
    "baseUrl": "http://coverhound.com/api",
  },
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions remoteClear() clears the resource 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {},
  },
}
`;

exports[`reducer actions remoteCreate() increments isCreating 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 1,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions remoteCreate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions remoteDelete() increments isDeleting and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 1,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`reducer actions remoteDelete() removes the resources from the store 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
    },
  },
}
`;

exports[`reducer actions remoteHydrate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions remoteRead() increments isReading 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 1,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions remoteRead() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions remoteUpdate() increments isUpdating and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 1,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`reducer actions remoteUpdate() updates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer initialState has an empty resources object 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;
